File: 15-3sum/15-3sum.java
import java.util.List;
import java.util.LinkedList;
import java.util.HashMap;
import java.util.Arrays;
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> out = new LinkedList<>();

        Arrays.sort(nums);
        for(int start = 0; start < nums.length ; start++){
            if(start > 0 && nums[start] == nums[start-1]){
                continue; 
            }else{
                var l = start+1;
                var r = nums.length-1;
                while(r > l){
                    var sum = nums[start] + nums[r] + nums[l];
                    if(sum > 0){
                        r--;
                    }else if(sum < 0){
                        l++;
                    }else{
                        out.add(List.of(nums[start], nums[r], nums[l]));
                        l++;
                        while(nums[l] == nums[l-1] && l < r){l++;}
                    }
                }
            }
        }
        return out;
    }
}

File: 15-3sum/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int l = 0, r = 0, sum = 0;
        for(int i = 0; i < nums.size(); i++){
            if(i > 0 and nums[i] == nums[i-1]){
                continue;
            }else{
                l = i+1;
                r = nums.size()-1;
                while(l < nums.size() and r >= 0 and r > l){
                    sum = nums[i] + nums[l] + nums[r];
                    if(sum == 0){
                        ans.push_back({nums[i], nums[l], nums[r]});
                        while(l + 1 < nums.size() and l+1 < r and nums[l] == nums[l+1]){
                            l++;
                        }
                        l++;
                    }else if(sum > 0){
                        r--;
                    }else{
                        l++;
                    }
                }
            }
        }
        return ans;
    }
};